[Serve] Add status code to retry when request timed out (#46527)
ray-project/ray: doc/source/serve/advanced-guides/performance.md

By default, Serve lets client HTTP requests run to completion no matter how long they take. However, slow requests could bottleneck the replica processing, blocking other requests that are waiting. Set an end-to-end timeout, so slow requests can be terminated and retried.

> You can set an end-to-end timeout for HTTP requests by setting the `request_timeout_s`

Reviewer: ```suggestion
You can set an end-to-end timeout for HTTP requests by setting the `request_timeout_s` parameter
```

in the `http_options` field of the Serve config. HTTP Proxies wait for that many
seconds before terminating an HTTP request. This config is global to your Ray cluster,

> and it can't be updated during runtime. Use [client-side retries](serve-best-practices-http-requests)

Reviewer: ```suggestion
and you can't update it during runtime. Use [client-side retries](serve-best-practices-http-requests)
```

By default, Serve lets client HTTP requests run to completion no matter how long they take. However, slow requests could bottleneck the replica processing, blocking other requests that are waiting. Set an end-to-end timeout, so slow requests can be terminated and retried.

You can set an end-to-end timeout for HTTP requests by setting the `request_timeout_s`
in the `http_options` field of the Serve config. HTTP Proxies wait for that many
seconds before terminating an HTTP request. This config is global to your Ray cluster,
and it can't be updated during runtime. Use [client-side retries](serve-best-practices-http-requests)
to retry requests that time out due to transient failures.

> :::{note}
> Serve returns response with status code `408` when the request timed out. Clients can
> handle this status code and retry the request.
> :::

Reviewer 2: ```suggestion
:::{note}
Serve returns a response with status code `408` when a request times out. Clients can retry when they receive this `408` response.
:::
```

Author: Yea, I don't think the client have to retry every 408. And good point about retrying long running processing likely to timed out again. But the goal is really to let the user know what status code to expect and handle accordingly.
